orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/util/ArtifactResolver.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2020 Google, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.pipeline.util;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.lang.String.format;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.netflix.spinnaker.kork.artifacts.model.Artifact;
+import com.netflix.spinnaker.kork.artifacts.model.ExpectedArtifact;
+import com.netflix.spinnaker.kork.web.exceptions.InvalidRequestException;
+import java.util.Optional;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.annotation.ParametersAreNonnullByDefault;
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+/** This class handles resolving a collection of {@link ExpectedArtifact} instances. */
+@ParametersAreNonnullByDefault
+public final class ArtifactResolver {
+  @Nonnull private final ImmutableList<Artifact> currentArtifacts;
+  @Nonnull private final ImmutableList<Artifact> priorArtifacts;
+  private final boolean requireUniqueMatches;
+
+  private ArtifactResolver(
+      @Nullable Iterable<Artifact> currentArtifacts,
+      @Nullable Iterable<Artifact> priorArtifacts,
+      boolean requireUniqueMatches) {
+    this.currentArtifacts =
+        currentArtifacts == null ? ImmutableList.of() : ImmutableList.copyOf(currentArtifacts);
+    this.priorArtifacts =
+        priorArtifacts == null ? ImmutableList.of() : ImmutableList.copyOf(priorArtifacts);
+    this.requireUniqueMatches = requireUniqueMatches;
+  }
+
+  /**
+   * Returns an instance of an {@link ArtifactResolver} that resolves against the supplied current
+   * artifacts, and prior artifacts.
+   *
+   * @param currentArtifacts The current artifacts to consider when resolving expected artifacts
+   * @param priorArtifacts The prior artifacts to consider when resolving expected artifacts
+   * @param requireUniqueMatches Whether the resolver should require that each expected artifact
+   *     matches at most one artifact.
+   * @return An instance of {@link ArtifactResolver}
+   */
+  public static ArtifactResolver getInstance(
+      @Nullable Iterable<Artifact> currentArtifacts,
```

Comment:
This is personal preference, so feel free to ignore it, but I think `@Nullable` parameters should be used sparingly, especially for collection parameters. `currentArtifacts` is never null (in production code anyway) and clients could just pass `ImmutableSet.of()` instead of `null` to `priorArtifacts`...  (Or just add an overload that removes the `priorArtifacts` parameter entirely, which is arguably a better move anyway)

To me, nullable types have pretty serious downsides, so I try to avoid passing across API boundaries unless there's a compelling reason, but I don't really see one here...

Reply:
My thought here was that it would be more flexible for API consumers if the API accepted and did something sensible with null values (but never returned null values)...but I can see how it might just encourage consumers to avoid thinking about what nulls mean.  I've updated the methods in the class to no longer accept null parameters, and added an overload to handle the case of `priorArtifacts` being null.